File: android/app/src/main/java/com/voicenote/app/core/audio/PcmUtil.kt
package com.voicenote.app.core.audio

/** PCM 音频数据转换工具 */
object PcmUtil {

    /**
     * 将 16-bit 小端序 PCM 字节数组转换为归一化浮点数组 [-1, 1]。
     * 供 ASR 推理与实时音量计算复用。
     */
    fun convertPCMToFloats(pcmData: ByteArray): FloatArray {
        val sampleCount = pcmData.size / 2
        val floats = FloatArray(sampleCount)
        var offset = 0
        for (i in 0 until sampleCount) {
            val sample = ((pcmData[offset + 1].toInt() shl 8) or
                          (pcmData[offset].toInt() and 0xFF)).toShort()
            floats[i] = sample.toFloat() / 32768.0f
            offset += 2
        }
        return floats
    }
}

File: desktop/internal/audio/pcm_util.go
// Package audio provides PCM audio capture and playback.
package audio

// ConvertPCMToFloats converts 16-bit little-endian PCM bytes to normalized
// float32 samples in [-1, 1]. It is shared by ASR inference and audio-level
// computation.
func ConvertPCMToFloats(pcm []byte) []float32 {
	sampleCount := len(pcm) / 2
	floats := make([]float32, sampleCount)
	offset := 0
	for i := 0; i < sampleCount; i++ {
		sample := int16(uint16(pcm[offset]) | uint16(pcm[offset+1])<<8)
		floats[i] = float32(sample) / 32768.0
		offset += 2
	}
	return floats
}

// ConvertFloatsToPCM converts normalized float32 samples to 16-bit
// little-endian PCM bytes.
func ConvertFloatsToPCM(samples []float32) []byte {
	buf := make([]byte, len(samples)*2)
	for i, s := range samples {
		if s > 1.0 {
			s = 1.0
		} else if s < -1.0 {
			s = -1.0
		}
		v := int16(s * 32767.0)
		buf[i*2] = byte(v)
		buf[i*2+1] = byte(uint16(v) >> 8)
	}
	return buf
}
